nexus/db-queries/src/db/datastore/region_replacement.rs

```diff
@@ -0,0 +1,886 @@
+        use db::schema::region_replacement::dsl;
+
+        dsl::region_replacement
+            .filter(
+                dsl::replacement_state.eq(RegionReplacementState::Requested),
+            )
+            .get_results_async::<RegionReplacement>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Return region replacement requests that are in state `Running` with no
+    /// currently operating saga. These need to be checked on or driven forward.
+    pub async fn get_running_region_replacements(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<Vec<RegionReplacement>, Error> {
+        use db::schema::region_replacement::dsl;
+
+        dsl::region_replacement
+            .filter(dsl::replacement_state.eq(RegionReplacementState::Running))
+            .filter(dsl::operating_saga_id.is_null())
+            .get_results_async::<RegionReplacement>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Return region replacement requests that are in state `ReplacementDone`
+    /// with no currently operating saga. These need to be completed.
+    pub async fn get_done_region_replacements(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<Vec<RegionReplacement>, Error> {
+        use db::schema::region_replacement::dsl;
+
+        dsl::region_replacement
+            .filter(
+                dsl::replacement_state
+                    .eq(RegionReplacementState::ReplacementDone),
+            )
+            .filter(dsl::operating_saga_id.is_null())
+            .get_results_async::<RegionReplacement>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Transition a RegionReplacement record from Requested to Allocating,
+    /// setting a unique id at the same time.
+    pub async fn set_region_replacement_allocating(
+        &self,
+        opctx: &OpContext,
+        region_replacement_id: Uuid,
+        operating_saga_id: Uuid,
+    ) -> Result<(), Error> {
+        use db::schema::region_replacement::dsl;
+        let updated = diesel::update(dsl::region_replacement)
+            .filter(dsl::id.eq(region_replacement_id))
+            .filter(
+                dsl::replacement_state.eq(RegionReplacementState::Requested),
+            )
+            .set((
+                dsl::replacement_state.eq(RegionReplacementState::Allocating),
+                dsl::operating_saga_id.eq(operating_saga_id),
+            ))
+            .check_if_exists::<RegionReplacement>(region_replacement_id)
+            .execute_and_check(&*self.pool_connection_authorized(opctx).await?)
```

Comment:
1. I'm confused, there's no use of `pool_connection_unauthorized` in this file
2. Yeah, in this case it works at the time of getting a pool connection. If this were working with user-visible objects (like `Disk`), then there could be additional permission checks.

Reply:
1.  I mean this method calls:
```
.execute_and_check(&*self.pool_connection_authorized(opctx).await?)
```
But, other methods above it don't have this line.  Is that because this method
is doing a call to `set()` and the other methods do not?

2. It was really the placement of the `..execute_and_check()` in the list of operations
that happen.  Like I wondered if this call was checking that the caller has the proper
permissions to execute the previous statements and if it is a permission check, that
it has to come at the end.

Reply:
ah, I think I get what you're asking now: `execute_and_check` is a CTE that is basically an UPDATE but it returns some more information, like if the record existed previously or not. It's not related to the opctx or permissions structure that we have set up.